File: claude_max_proxy/proxy.py
#!/usr/bin/env python3
"""
claude-max-proxy — Drop-in fix for OpenClaw when Anthropic blocks your Max subscription.

Reads your openclaw.json, loads all agents and skills, and exposes an HTTP
webhook that the OpenClaw gateway calls for agent execution. Works with all
OpenClaw channels (Telegram, WhatsApp, Slack, Discord, Signal, iMessage, etc.)

Usage:
    claude-max-proxy                          # auto-detect ~/.openclaw/openclaw.json
    claude-max-proxy /path/to/openclaw.json   # explicit config path
    claude-max-proxy --port 8080              # custom port
"""

import argparse
import asyncio
import json
import os
import sys
from datetime import datetime
from pathlib import Path
from http import HTTPStatus

from aiohttp import web
from claude_agent_sdk import query, ClaudeAgentOptions, ResultMessage


# ---------------------------------------------------------------------------
# Config loader
# ---------------------------------------------------------------------------


def find_config(cli_path: str = None) -> Path:
    """Find openclaw.json — CLI arg, env var, or default paths."""
    if cli_path:
        return Path(cli_path)
    if env := os.environ.get("OPENCLAW_CONFIG"):
        return Path(env)
    for candidate in [
        Path.home() / ".openclaw" / "openclaw.json",
        Path("/root/.openclaw/openclaw.json"),
    ]:
        if candidate.exists():
            return candidate
    print("ERROR: openclaw.json not found. Pass the path as argument or set OPENCLAW_CONFIG.")
    sys.exit(1)


def load_config(config_path: Path) -> dict:
    with open(config_path) as f:
        return json.load(f)


# ---------------------------------------------------------------------------
# Agent prompt builder
# ---------------------------------------------------------------------------

PROMPT_FILES = ["IDENTITY.md", "SOUL.md", "AGENTS.md", "TOOLS.md", "USER.md", "HEARTBEAT.md"]


def load_agent_prompt(workspace: str, skills_dir: str = None) -> str:
    """Build system prompt from an agent's workspace directory."""
    parts = []

    for filename in PROMPT_FILES:
        filepath = os.path.join(workspace, filename)
        if os.path.exists(filepath):
            parts.append(open(filepath).read().strip())

    if skills_dir and os.path.isdir(skills_dir):
        for skill_name in sorted(os.listdir(skills_dir)):
            skill_md = os.path.join(skills_dir, skill_name, "SKILL.md")
            if os.path.exists(skill_md):
                parts.append(f"## Skill: {skill_name}\n\n{open(skill_md).read().strip()}")

    for f in sorted(os.listdir(workspace)):
        if f.endswith(".json"):
            filepath = os.path.join(workspace, f)
            try:
                content = open(filepath).read().strip()
                parts.append(f"## Current state: {f}\n\n```json\n{content}\n```")
            except Exception:
                pass

    return "\n\n---\n\n".join(parts)


# ---------------------------------------------------------------------------
# Agent SDK execution
# ---------------------------------------------------------------------------


async def run_agent(
    system_prompt: str,
    message: str,
    model: str = "claude-sonnet-4-6",
    workspace: str = None,
) -> str:
    """Send a message to an agent via the Claude Agent SDK."""
    result = ""
    try:
        async for msg in query(
            prompt=message,
            options=ClaudeAgentOptions(
                system_prompt=system_prompt,
                allowed_tools=["WebSearch", "WebFetch", "Read", "Write", "Bash"],
                model=model,
                max_turns=15,
                **({"cwd": workspace} if workspace else {}),
            ),
        ):
            if isinstance(msg, ResultMessage):
                result = msg.result
    except Exception as e:
        result = f"[proxy error] {e}"
    return result


# ---------------------------------------------------------------------------
# Agent registry
# ---------------------------------------------------------------------------


class AgentRegistry:
    """Holds resolved agent configs loaded from openclaw.json."""

    def __init__(self, config: dict):
        self.config = config
        self.agents: dict[str, dict] = {}
        self._resolve()

    def _resolve(self):
        agents_list = self.config.get("agents", {}).get("list", [])
        default_model = self.config.get("agents", {}).get("defaultModel", "anthropic/claude-sonnet-4-6")
        skills_dir = os.path.expanduser("~/.openclaw/skills")

        for agent in agents_list:
            agent_id = agent["id"]
            workspace = agent.get("workspace", "")
            if not workspace or not os.path.isdir(workspace):
                log(agent_id, f"workspace not found: {workspace}, skipping")
                continue

            model_raw = agent.get("model") or default_model
            model = model_raw.replace("anthropic/", "")
            identity = agent.get("identity", {})
            display_name = identity.get("name", agent_id)
            emoji = identity.get("emoji", "")

            self.agents[agent_id] = {
                "agent_id": agent_id,
                "display_name": f"{emoji} {display_name}".strip(),
                "workspace": workspace,
                "model": model,
                "skills_dir": skills_dir,
            }


# ---------------------------------------------------------------------------
# HTTP webhook server
# ---------------------------------------------------------------------------


async def handle_agent(request: web.Request) -> web.Response:
    """
    POST /v1/agent/{agent_id}
    Body: {"message": "..."}
    Response: {"reply": "...", "agent_id": "...", "model": "..."}
    """
    registry: AgentRegistry = request.app["registry"]
    agent_id = request.match_info["agent_id"]

    agent = registry.agents.get(agent_id)
    if not agent:
        return web.json_response(
            {"error": f"Unknown agent: {agent_id}"},
            status=HTTPStatus.NOT_FOUND,
        )

    try:
        body = await request.json()
    except Exception:
        return web.json_response(
            {"error": "Invalid JSON body"},
            status=HTTPStatus.BAD_REQUEST,
        )

    message = body.get("message", "")
    if not message:
        return web.json_response(
            {"error": "Missing 'message' field"},
            status=HTTPStatus.BAD_REQUEST,
        )

    log(agent["display_name"], f"← {message[:80]}")

    prompt = load_agent_prompt(agent["workspace"], agent["skills_dir"])
    reply = await run_agent(prompt, message, model=agent["model"], workspace=agent["workspace"])

    log(agent["display_name"], f"→ {reply[:80]}")

    return web.json_response({
        "reply": reply,
        "agent_id": agent_id,
        "model": agent["model"],
    })


async def handle_list_agents(request: web.Request) -> web.Response:
    """GET /v1/agents — list all available agents."""
    registry: AgentRegistry = request.app["registry"]
    return web.json_response({
        "agents": [
            {"id": a["agent_id"], "name": a["display_name"], "model": a["model"]}
            for a in registry.agents.values()
        ]
    })


async def handle_health(request: web.Request) -> web.Response:
    """GET /health"""
    return web.json_response({"status": "ok"})


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


def log(agent: str, msg: str):
    ts = datetime.now().strftime("%H:%M:%S")
    print(f"[{ts}] [{agent}] {msg}")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="claude-max-proxy for OpenClaw")
    parser.add_argument("config", nargs="?", help="Path to openclaw.json")
    parser.add_argument("--port", type=int, default=3777, help="Server port (default: 3777)")
    parser.add_argument("--host", default="0.0.0.0", help="Server host (default: 0.0.0.0)")
    return parser.parse_args()


async def main():
    args = parse_args()

    config_path = find_config(args.config)
    log("proxy", f"loading {config_path}")
    config = load_config(config_path)
    registry = AgentRegistry(config)

    if not registry.agents:
        print("ERROR: No agents found in config.")
        sys.exit(1)

    print(f"\n  claude-max-proxy — {len(registry.agents)} agents loaded\n")
    for a in registry.agents.values():
        print(f"  {a['display_name']:20s}  model={a['model']}")
    print()

    app = web.Application()
    app["registry"] = registry
    app.router.add_post("/v1/agent/{agent_id}", handle_agent)
    app.router.add_get("/v1/agents", handle_list_agents)
    app.router.add_get("/health", handle_health)

    runner = web.AppRunner(app)
    await runner.setup()
    site = web.TCPSite(runner, args.host, args.port)
    await site.start()

    log("proxy", f"listening on http://{args.host}:{args.port}")
    log("proxy", f"  POST /v1/agent/{{agent_id}}")
    log("proxy", f"  GET  /v1/agents")
    log("proxy", f"  GET  /health")

    await asyncio.Event().wait()


def cli():
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\nStopped.")


if __name__ == "__main__":
    cli()

File: src/proxy.ts
#!/usr/bin/env node

/**
 * claude-max-proxy — Drop-in fix for OpenClaw when Anthropic blocks your Max subscription.
 *
 * Reads your openclaw.json, loads all agents and skills, and exposes an HTTP
 * webhook that the OpenClaw gateway calls for agent execution. Works with all
 * OpenClaw channels (Telegram, WhatsApp, Slack, Discord, Signal, iMessage, etc.)
 *
 * Usage:
 *     npx claude-max-proxy                          # auto-detect ~/.openclaw/openclaw.json
 *     npx claude-max-proxy /path/to/openclaw.json   # explicit config path
 *     npx claude-max-proxy --port 8080              # custom port
 */

import { createServer, type IncomingMessage, type ServerResponse } from "node:http";
import { readFileSync, existsSync, readdirSync } from "node:fs";
import { join, resolve } from "node:path";
import { homedir } from "node:os";
import { query, type SDKResultMessage } from "@anthropic-ai/claude-agent-sdk";

// ---------------------------------------------------------------------------
// Config loader
// ---------------------------------------------------------------------------

const PROMPT_FILES = [
  "IDENTITY.md",
  "SOUL.md",
  "AGENTS.md",
  "TOOLS.md",
  "USER.md",
  "HEARTBEAT.md",
];

interface AgentConfig {
  id: string;
  workspace?: string;
  model?: string;
  identity?: { name?: string; emoji?: string };
}

interface OpenClawConfig {
  agents?: {
    list?: AgentConfig[];
    defaultModel?: string;
  };
}

interface ResolvedAgent {
  agentId: string;
  displayName: string;
  workspace: string;
  model: string;
  skillsDir: string;
}

function findConfig(cliPath?: string): string {
  if (cliPath) return resolve(cliPath);

  const envPath = process.env["OPENCLAW_CONFIG"];
  if (envPath) return resolve(envPath);

  const candidates = [
    join(homedir(), ".openclaw", "openclaw.json"),
    "/root/.openclaw/openclaw.json",
  ];
  for (const candidate of candidates) {
    if (existsSync(candidate)) return candidate;
  }

  console.error(
    "ERROR: openclaw.json not found. Pass the path as argument or set OPENCLAW_CONFIG."
  );
  process.exit(1);
}

function loadConfig(configPath: string): OpenClawConfig {
  return JSON.parse(readFileSync(configPath, "utf-8"));
}

// ---------------------------------------------------------------------------
// Agent prompt builder
// ---------------------------------------------------------------------------

function loadAgentPrompt(workspace: string, skillsDir?: string): string {
  const parts: string[] = [];

  for (const filename of PROMPT_FILES) {
    const filepath = join(workspace, filename);
    if (existsSync(filepath)) {
      parts.push(readFileSync(filepath, "utf-8").trim());
    }
  }

  if (skillsDir && existsSync(skillsDir)) {
    for (const skillName of readdirSync(skillsDir).sort()) {
      const skillMd = join(skillsDir, skillName, "SKILL.md");
      if (existsSync(skillMd)) {
        parts.push(
          `## Skill: ${skillName}\n\n${readFileSync(skillMd, "utf-8").trim()}`
        );
      }
    }
  }

  for (const f of readdirSync(workspace).sort()) {
    if (f.endsWith(".json")) {
      try {
        const content = readFileSync(join(workspace, f), "utf-8").trim();
        parts.push(`## Current state: ${f}\n\n\`\`\`json\n${content}\n\`\`\``);
      } catch {
        // skip unreadable files
      }
    }
  }

  return parts.join("\n\n---\n\n");
}

// ---------------------------------------------------------------------------
// Agent SDK execution
// ---------------------------------------------------------------------------

async function runAgent(
  systemPrompt: string,
  message: string,
  model: string = "claude-sonnet-4-6",
  workspace?: string
): Promise<string> {
  let result = "";
  try {
    for await (const msg of query({
      prompt: message,
      options: {
        systemPrompt,
        allowedTools: ["WebSearch", "WebFetch", "Read", "Write", "Bash"],
        model,
        maxTurns: 15,
        permissionMode: "bypassPermissions",
        allowDangerouslySkipPermissions: true,
        ...(workspace ? { cwd: workspace } : {}),
      },
    })) {
      if (msg.type === "result" && msg.subtype === "success") {
        result = (msg as SDKResultMessage & { result: string }).result;
      }
    }
  } catch (e) {
    result = `[proxy error] ${e}`;
  }
  return result;
}

// ---------------------------------------------------------------------------
// Agent registry
// ---------------------------------------------------------------------------

function resolveAgents(config: OpenClawConfig): Map<string, ResolvedAgent> {
  const agents = new Map<string, ResolvedAgent>();
  const agentsList = config.agents?.list ?? [];
  const defaultModel =
    config.agents?.defaultModel ?? "anthropic/claude-sonnet-4-6";
  const skillsDir = join(homedir(), ".openclaw", "skills");

  for (const agent of agentsList) {
    const workspace = agent.workspace ?? "";
    if (!workspace || !existsSync(workspace)) {
      log(agent.id, `workspace not found: ${workspace}, skipping`);
      continue;
    }

    const model = (agent.model ?? defaultModel).replace("anthropic/", "");
    const displayName =
      `${agent.identity?.emoji ?? ""} ${agent.identity?.name ?? agent.id}`.trim();

    agents.set(agent.id, {
      agentId: agent.id,
      displayName,
      workspace,
      model,
      skillsDir,
    });
  }

  return agents;
}

// ---------------------------------------------------------------------------
// HTTP server
// ---------------------------------------------------------------------------

function log(agent: string, msg: string): void {
  const ts = new Date().toTimeString().slice(0, 8);
  console.log(`[${ts}] [${agent}] ${msg}`);
}

function jsonResponse(
  res: ServerResponse,
  status: number,
  body: unknown
): void {
  res.writeHead(status, { "Content-Type": "application/json" });
  res.end(JSON.stringify(body));
}

async function readBody(req: IncomingMessage): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of req) {
    chunks.push(chunk as Buffer);
  }
  return Buffer.concat(chunks).toString("utf-8");
}

function parseArgs(): { configPath?: string; port: number; host: string } {
  const args = process.argv.slice(2);
  let configPath: string | undefined;
  let port = 3777;
  let host = "0.0.0.0";

  for (let i = 0; i < args.length; i++) {
    const arg = args[i]!;
    if (arg === "--port" && args[i + 1]) {
      port = parseInt(args[++i]!, 10);
    } else if (arg === "--host" && args[i + 1]) {
      host = args[++i]!;
    } else if (!arg.startsWith("-")) {
      configPath = arg;
    }
  }

  return { configPath, port, host };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const { configPath, port, host } = parseArgs();

  const resolvedConfigPath = findConfig(configPath);
  log("proxy", `loading ${resolvedConfigPath}`);
  const config = loadConfig(resolvedConfigPath);
  const agents = resolveAgents(config);

  if (agents.size === 0) {
    console.error("ERROR: No agents found in config.");
    process.exit(1);
  }

  console.log(`\n  claude-max-proxy — ${agents.size} agents loaded\n`);
  for (const a of agents.values()) {
    console.log(`  ${a.displayName.padEnd(20)}  model=${a.model}`);
  }
  console.log();

  const server = createServer(async (req, res) => {
    const url = new URL(req.url ?? "/", `http://${req.headers.host}`);

    // GET /health
    if (req.method === "GET" && url.pathname === "/health") {
      jsonResponse(res, 200, { status: "ok" });
      return;
    }

    // GET /v1/agents
    if (req.method === "GET" && url.pathname === "/v1/agents") {
      jsonResponse(res, 200, {
        agents: [...agents.values()].map((a) => ({
          id: a.agentId,
          name: a.displayName,
          model: a.model,
        })),
      });
      return;
    }

    // POST /v1/agent/:agent_id
    const agentMatch = url.pathname.match(/^\/v1\/agent\/([^/]+)$/);
    if (req.method === "POST" && agentMatch) {
      const agentId = agentMatch[1]!;
      const agent = agents.get(agentId);

      if (!agent) {
        jsonResponse(res, 404, { error: `Unknown agent: ${agentId}` });
        return;
      }

      let body: { message?: string };
      try {
        body = JSON.parse(await readBody(req));
      } catch {
        jsonResponse(res, 400, { error: "Invalid JSON body" });
        return;
      }

      const message = body.message ?? "";
      if (!message) {
        jsonResponse(res, 400, { error: "Missing 'message' field" });
        return;
      }

      log(agent.displayName, `← ${message.slice(0, 80)}`);

      const prompt = loadAgentPrompt(agent.workspace, agent.skillsDir);
      const reply = await runAgent(
        prompt,
        message,
        agent.model,
        agent.workspace
      );

      log(agent.displayName, `→ ${reply.slice(0, 80)}`);

      jsonResponse(res, 200, {
        reply,
        agent_id: agentId,
        model: agent.model,
      });
      return;
    }

    jsonResponse(res, 404, { error: "Not found" });
  });

  server.listen(port, host, () => {
    log("proxy", `listening on http://${host}:${port}`);
    log("proxy", `  POST /v1/agent/{agent_id}`);
    log("proxy", `  GET  /v1/agents`);
    log("proxy", `  GET  /health`);
  });
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
